### ISSUE TITLE
Reverse proxy support

### ISSUE THREAD

**Issue by PengZheng** [civetweb/civetweb#1131]
Is there a plan to add builtin support for reverse proxy (especially forwarding tcp connection to unix domain socket)? 
It will help migration from a existing embedded server to civetweb.
libwebsockets does support this, but its server API is not very easy to use.

**Comment by bel2125** [comment]
Actually I built something like that using CivetWeb by writing some small handler functions in C.
When starting from an example code, this is less than one day work to program.
I haven't put it into the main civetweb C code, since this is quite special - definitely not required for everyone.
I could have "cleaned" the code a bit and added it as an example if it is useful for several people.

In standard, CivetWeb allows to use domain sockets (in addition to tcp) to receive http requests.
It can be activated at build time using a compiler flag.

But you want the following: 
receive a HTTP/tcp request, forward the same request to a domain socket, wait for the response from the domain socket and forward it to the HTTP/tcp connection. You only forward the HTTP body data or also the header? I only forwarded the body data. In theory, you could also modify the data, bring it to a different format, add authentication support, etc.
I did not put it into the standard, because there are several options to do it and making them "configurable" would be too much - they are "programmable" by specifying what you want in the callback.
If your demands are similar to what I have done before, I might be able to extract the more common requirements into a simple piece of example code.

**Comment by PengZheng** [comment]
> You only forward the HTTP body data or also the header?
Both header (probably with some modification) and body should be forwarded. 
The route of incoming request should be determined by some user provided hooks, which can choose to check header only, header+part of the body, or header+full body. 

The biggest problem I see with the handler approach you suggest is performance. The thread pool is a very limited source. 
If the reverse proxy could be all done within a single thread, the situation will be very ideal.

**Comment by bel2125** [comment]
CivetWeb by design uses a thread pool, to allow using blocking handler functions.
Using one single thread would mean all your functions in the handler must access data that is always ready or available very fast, because if you need to wait for the data ... for example if you need to access it from some remote location or do a lot of calculations ... you will also block all other clients. If you enable connection keep-alive, one thread can handle multiple requests in one thread within the same connection. But a second connection from another client will be processed by another thread. This way one request from one connection cannot fully block another client connection.

**Comment by PengZheng** [comment]
> CivetWeb by design uses a thread pool, to allow using blocking handler functions.

Yes, that's why CivetWeb's API is very easy to use. Blocking handlers should be dispatched to thread pool.

What I have in mind is a combination of thread pool and non-blocking event loop, like:
https://github.com/warmcat/libwebsockets/tree/main/lib/misc/threadpool

The main feature of this design is that the thread pool and event loop can sync with each other, unlike libuv's thread pool design.
For blocking handler, its interface does not have to be changed, and is still dispatched to the thread pool. 
While for reverse proxy, it can be handled entirely in the event-loop.

https://github.com/warmcat/libwebsockets/blob/main/minimal-examples-lowlevel/ws-server/minimal-ws-server-threadpool/protocol_lws_minimal_threadpool.c

The above example does not achieve the simplicity of Civetweb, but is way easier to program than event-loop's callback hell.

Setting up a separate reverse proxy is one way to help business migration. But in that case, both old and new web service suffer from the cost of forwarding. For application like video streaming, it's unacceptable for resource-constrained device. The migration can last for a long time. If Civetweb have builtin support of reverse proxy, we can migrate such critical applications first to Civetweb, leaving other services untouched.

I fully understand implementing efficient reverse proxy may require fundamental design change. Feel free to close it if it won't be planned in the future.

**Comment by rlb567** [comment]
I am also interested in reverse proxy support, and yeah I agree this is somewhat a niche requirement that not everyone will need.  Any chance of getting the example code of how you've done it?  I don't care how clean it is or anything, more just as guidance to how/where you've done it.   In my case I want it to put haproxy headers on for the downstream web server.  https://www.haproxy.org/download/1.8/doc/proxy-protocol.txt   Anything that gives me a jump start to implementing this myself is helpful.
